components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/FxaAccountManager.kt

```diff
@@ -0,0 +1,423 @@
+                    when (event) {
+                        FetchProfile -> AccountState.AuthenticatedNoProfile
+                        FetchedProfile -> AccountState.AuthenticatedWithProfile
+                        FailedToFetchProfile -> AccountState.AuthenticatedNoProfile
+                        Logout -> AccountState.NotAuthenticated
+                        else -> null
+                    }
+                }
+                AccountState.AuthenticatedWithProfile -> {
+                    when (event) {
+                        Logout -> AccountState.NotAuthenticated
+                        else -> null
+                    }
+                }
+            }
+        }
+    }
+
+    private val exceptionHandler = CoroutineExceptionHandler { _, e ->
+        // TODO emit an "error" event to the state machine, and have individual states deal with it as appropriate.
+        Log.log(
+            Log.Priority.ERROR,
+            message = "Unexpected error occurred during Firefox Account state transitions. Current state: $state.",
+            throwable = e,
+            tag = logTag
+        )
+    }
+
+    private var job = SupervisorJob()
+    // We want a single-threaded execution model for our account-related "actions" (state machine side-effects).
+    // That is, we want to ensure a sequential execution flow, but on a background thread.
+    private val coroutineContext: CoroutineContext
+        get() = Executors.newSingleThreadExecutor().asCoroutineDispatcher() + job + exceptionHandler
+
+    // 'account' is initialized during processing of an 'Init' event.
+    // Note on threading: we use a single-threaded executor, so there's no concurrent access possible.
+    // However, that executor doesn't guarantee that it'll always use the same thread, and so vars
+    // are marked as volatile for across-thread visibility. Similarly, event queue uses a concurrent
+    // list, although that's probably an overkill.
+    @Volatile private lateinit var account: FirefoxAccountShaped
+    @Volatile private var profile: Profile? = null
+    @Volatile private var state = AccountState.Start
+    private val eventQueue = ConcurrentLinkedQueue<AccountEvent>()
+
+    /**
+     * Call this after registering your observers, and before interacting with this class.
+     */
+    fun init(): Deferred<Unit> {
+        return processQueue(Init)
+    }
+
+    fun authenticatedAccount(): FirefoxAccountShaped? {
+        return when (state) {
+            AccountState.AuthenticatedWithProfile,
+            AccountState.AuthenticatedNoProfile -> account
+            else -> null
+        }
+    }
+
+    fun accountProfile(): Profile? {
+        return when (state) {
+            AccountState.AuthenticatedWithProfile -> profile
+            else -> null
+        }
+    }
+
+    fun beginAuthentication(): Deferred<String> {
+        val deferredAuthUrl: CompletableDeferred<String> = CompletableDeferred()
+
+        oauth.register(object : OAuthObserver {
```

Comment:
This looks a bit weird to be honest, could we get rid of the `OAuthObserver` altogether? @csadilek told me it's more like a leftover implementation detail

Reply:
I'm not too happy with this, but it's a straightforward way to implement this, and both `oauth` and `OAuthObserver` are private to this class so the outside world know about them.